syncer/dispatcher.go

```diff
@@ -0,0 +1,159 @@
+package syncer
+
+import (
+	"container/heap"
+	"errors"
+	"sync"
+
+	"github.com/filecoin-project/go-filecoin/types"
+)
+
+var errBadPush = errors.New("a programmer is pushing the wrong type to a TargetQueue")
+var errBadPop = errors.New("a programmer is not checking targetQueue length before popping")
+
+// NewDispatcher creates a new syncing dispatcher.
+func NewDispatcher() *Dispatcher {
+
+	return &Dispatcher{
+		targetSet: make(map[string]struct{}),
+		targetQ:   NewTargetQueue(),
+	}
+}
+
+// Dispatcher executes syncing requests
+type Dispatcher struct {
+	// The dispatcher maintains a targeting system for determining the
+	// current best syncing target
+	// targetMu protects the targeting system
+	targetMu sync.Mutex
+	// targetSet tracks all tipsetkeys currently being targeted to prevent
+	// pushing duplicates to the target queue
+	targetSet map[string]struct{}
+	// targetQ is a priority queue of target tipsets
+	targetQ *TargetQueue
+}
+
+// ReceiveHello handles chain information from bootstrap peers.
+func (d *Dispatcher) ReceiveHello(ci types.ChainInfo) error { return d.receive(ci) }
+
+// ReceiveOwnBlock handles chain info from a node's own mining system
+func (d *Dispatcher) ReceiveOwnBlock(ci types.ChainInfo) error { return d.receive(ci) }
+
+// ReceiveGossipBlock handles chain info from new blocks sent on pubsub
+func (d *Dispatcher) ReceiveGossipBlock(ci types.ChainInfo) error { return d.receive(ci) }
+
+func (d *Dispatcher) receive(ci types.ChainInfo) error {
+	d.targetMu.Lock()
+	defer d.targetMu.Unlock()
+
+	_, targeting := d.targetSet[ci.Head.String()]
+	if targeting {
+		// already tracking drop quickly
+		return nil
+	}
+	err := d.targetQ.Push(&SyncRequest{ChainInfo: ci})
+	if err != nil {
+		return err
+	}
+	d.targetSet[ci.Head.String()] = struct{}{}
+	return nil
+}
+
+// Start initialize the run loop that pops from the target queue and
+// dispatches syncRequests against the correct syncer.
+//
+// This method determines the default "business logic" of the syncing
+// subsystem.  It is responsible for interpreting information from the network
+// and using it to securely update the node's chain state.
+func (d *Dispatcher) Start() {
```

Comment:
This is a lot of promise with no delivery. Probably better to leave this out until it's at least partially implemented.